bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java

```diff
@@ -1066,5 +1059,73 @@ public String toString()
 
   }
 
+  private class ListenersNotifier implements Runnable
+  {
+    private boolean isMoreDataAvailable = false;
+
+    private synchronized void moreDataAvailable()
+    {
+      if (future == null) {
+        future = autoFlushExecutor.submit(this);
+      } else {
+        isMoreDataAvailable = true;
+      }
+    }
+
+    private boolean notifyListeners()
+    {
+      boolean doesAtLeastOneListenerHaveDataToSend = false;
+      for (DataListener dl : all_listeners) {
+        try {
+          doesAtLeastOneListenerHaveDataToSend |= dl.addedData(false);
+        } catch (RuntimeException e) {
+          logger.error("{}: removing DataListener {} due to exception", DataList.this, dl, e);
+          removeDataListener(dl);
+          break;
+        }
+      }
+      return doesAtLeastOneListenerHaveDataToSend;
+    }
+
+    private boolean checkIfListenersHaveDataToSendOnly()
+    {
+      synchronized (DataList.this) {
```

Comment:
It was originally synchronized to protect `future`, not the DataList. I refactored to avoid the synchronization.